### PR TITLE
memory_reduction

### PR DESCRIPTION
**Pull request by jposada202020**

### CODE REVIEW

**Comment by jposada202020** [comment]
Please feel free to suggest name changes, and comments @tannewt 

### Notes:

I have it working alone for the Qt-Py with a `py` extension
```python
Adafruit CircuitPython 6.1.0 on 2021-01-21; Adafruit QT Py M0 with samd21e18
>>> import dps310_simpletest
Temperature = 25.01 *C
Pressure = 1020.45 hPa

```

However when I added the code for the Full Featured class in the same file, I got a memory error.

So I decided to separate the files and include the light class in the ``__init__.py`` however also have a memory error. 

So two things here:
1. is the file size a factor here?
2. I know python while importing will parse the files, so even with putting the lighter version in the ``__init__`` it will parse the other files in the directory, and maybe the qt-py runs out of memory while parsing?


### Conclusion
The present change will have memory allocation error, if loaded with th py file, however it tested ok with the mpy files.

```python
Adafruit CircuitPython 6.1.0 on 2021-01-21; Adafruit QT Py M0 with samd21e18
>>> import dps310_simpletest
Temperature = 31.04 *C
Pressure = 1021.47 hPa


```

**Comment by tannewt** [comment]
> * is the file size a factor here?

If you get a memory error on import then use an mpy. It'll strip the white space and comments. Comments take space when parsing I believe. I think the resulting imported size is still the same.

To measure the memory footprint after import use `gc.mem_free()`.

> * I know python while importing will parse the files, so even with putting the lighter version in the `__init__` it will parse the other files in the directory, and maybe the qt-py runs out of memory while parsing?

It'll only parse the file that matches the import statement. That file may also import which would cause more files to be parsed.

**Comment by jposada202020** [comment]
Just to be clear, in our options here. Fail means= Memory Error - allocation Error

| Method           | Extension        | Test  |
|-----------------------------------------------------------------------------------------------|----------------------------------------|-----|
| Original Library                                                     | mpy         | Fail  |
| Customized File for QtPy only                                                | py         | Pass  |
|File with two classes, one for QTPy + DPS310_Featured                 |mpy       | Fail|
|library split in two `__init__` + DPS310_Featured    | py | Fail|
|library split in two `__init__` + DPS310_Featured    | mpy | Pass|

So according to this, we have two options. Two differents files, one for the Qtpy, or the library with the general methods in the `_init_.py` and only to be used for the qtpy as `mpy`

**Comment by tannewt** [comment]
@jposada202020 what is the difference between the qtpy version that works with py and the `__init__` that works as mpy? Does the `__init__` alone work as py too?

**Comment by jposada202020** [comment]
> @jposada202020 what is the difference between the qtpy version that works with py ?

I tailored the class, to have only basic configuration for the QTpy and nothing else. the ``__init__`` version has, basic configuration in the ``_init_`` and the full configuration class in a diferent py file. So basically the `__init__` version is the version included in the current PR. You could use the initi version apart and work with the QTPy alone, however when they are packaged like this they do not work, see explanation below.

> @jposada202020 Does the `__init__` alone work as py too?

No it does not, I tried but it does not, do not understand why, that is why I asking about the parsing in the import, I do not have an explanation on why it does not work. I am not importing the Class CV and other properties if I am only using the class in the init, however it does not work.

**Comment by jposada202020** [comment]
@tannewt it worked!!! 🛩️  works with py files. Tested in QTPY and Feather RP2040. Could you please propose a better name for the Featured class, I used Configurable, but I do not like it. Documentation is good, I add another example, so once you propose a better name I could polish and do the little details

**Comment by tannewt** [comment]
> @tannewt it worked!!! small_airplane works with py files. Tested in QTPY and Feather RP2040. Could you please propose a better name for the Featured class, I used Configurable, but I do not like it. Documentation is good, I add another example, so once you propose a better name I could polish and do the little details

How about `basic` and `advanced`? I don't think you need to call it `dps310_basic` either. `basic` is just fine because the package has `dps310` in it.

**Comment by jposada202020** [comment]
Good Idea, will do changes as suggested. I like the idea of basic and advanced... and have the basic file without a name

**Comment by jposada202020** [comment]
Test were done using different boards, and different versions of CircuitPython

## Tests done for the RP2040
1. Usage code inside the class for readthedocs --> OK
2. Code in `Readme.rst` --> OK
3. Three examples

```python
Adafruit CircuitPython 6.2.0 on 2021-04-05; Adafruit Feather RP2040 with rp2040
>>> import dps310_simpletest_advanced
Temperature = 26.55 *C
Pressure = 1025.00 hPa
```

```python
Adafruit CircuitPython 6.2.0 on 2021-04-05; Adafruit Feather RP2040 with rp2040
>>> import dps310_low_power_weather_station
Temperature = 26.46 *C
Pressure = 1025.00 hPa
```

```python
Adafruit CircuitPython 6.2.0 on 2021-04-05; Adafruit Feather RP2040 with rp2040
>>> import dps310_simpletest
Temperature = 26.37 *C
Pressure = 1025.02 hPa
```

## Tests done for the QTPY MO
Only the `simpletest` was done here as other class will not load because memory restrictions

```python
Adafruit CircuitPython 6.1.0 on 2021-01-21; Adafruit QT Py M0 with samd21e18
>>> import dps310_simpletest
Temperature = 26.55 *C
Pressure = 1025.01 hPa
```

**NOTE** 
### For I do not what reason I had to change the import order for the QTPY in the example code to avoid the memory Error. I let the designers explain that to me. Thanks


## Tests done for the Metro ESP32S2
```python
Adafruit CircuitPython 7.0.0-alpha.1-853-gcbe1aa74f on 2021-04-26; Adafruit Metro ESP32S2 with ESP32S2
>>> import dps310_simpletest
Temperature = 26.21 *C
Pressure = 1025.00 hPa
```

```python
Adafruit CircuitPython 7.0.0-alpha.1-853-gcbe1aa74f on 2021-04-26; Adafruit Metro ESP32S2 with ESP32S2
>>> import dps310_simpletest_advanced
Temperature = 26.28 *C
Pressure = 1024.97 hPa
```

```python
Adafruit CircuitPython 7.0.0-alpha.1-853-gcbe1aa74f on 2021-04-26; Adafruit Metro ESP32S2 with ESP32S2
>>> import dps310_low_power_weather_station
Temperature = 26.34 *C
Pressure = 1024.93 hPa
```

## Tests done for the  FeatherS2
```python
Adafruit CircuitPython 6.1.0-beta.2 on 2020-12-03; FeatherS2 with ESP32S2
>>> import dps310_simpletest
Temperature = 26.20 *C
Pressure = 1024.91 hPa
```

```python
Adafruit CircuitPython 6.1.0-beta.2 on 2020-12-03; FeatherS2 with ESP32S2
>>> import dps310_simpletest_advanced
Temperature = 26.22 *C
Pressure = 1024.89 hPa
```

```python
Adafruit CircuitPython 6.1.0-beta.2 on 2020-12-03; FeatherS2 with ESP32S2
>>> import dps310_low_power_weather_station
Temperature = 26.18 *C
Pressure = 1024.89 hPa
```

# NOTE: FOR RELEASING PURPOSES THIS IS A MAJOR CHANGE, IT WILL REQUIRE VERSION CHANGE

**Comment by tannewt** [comment]
> Good Idea, will do changes as suggested. I like the idea of basic and advanced... and have the basic file without a name

I think you should have `adafruit_dps310/basic.py` and `adafruit_dps310/advanced.py`. No need to have dps310 in the name twice.

**Comment by jposada202020** [comment]
Will do :) Thanks

**Comment by jposada202020** [comment]
Final test was done with RP2040 and Qtpy M0. the simpletest did not load with a py, however with the mpy files was ok

```python
Adafruit CircuitPython 6.2.0 on 2021-04-05; Adafruit QT Py M0 with samd21e18
>>> import dps310_simpletest
Temperature = 24.78 *C
Pressure = 1016.48 hPa

Temperature = 24.79 *C
Pressure = 1016.48 hPa
```

When extracting the basic file from the directory the memory error go away, and before changing the import order also correct the memory error, but when testing today this did not happened. To discuss in Discor. (was my two questions las friday :) )

**Comment by tannewt** [comment]
@jposada202020 want to do anything else on this?

**Comment by jposada202020** [comment]
@tannewt I think we are good to merge,however if you agree, could we wait until I do the BME280?, I want to see if I could learn/have some other elements that maybe could help here also.

**Comment by tannewt** [comment]
Up to you. :-)

File: README.rst
Comment (lesamouraipourpre) [inline README.rst]:
```suggestion
    from adafruit_dps310.basic import DPS310
```

File: README.rst
Comment (lesamouraipourpre) [inline README.rst]:
```suggestion
For boards with more memory available you could use the code present
```

File: README.rst
Comment (lesamouraipourpre) [inline README.rst]:
```suggestion
in ``adafruit_dps310/dps310_advanced.py``. For usage refer to ``dps310_simpletest_advanced.py``
```

File: adafruit_dps310/advanced.py
Comment (lesamouraipourpre) [inline adafruit_dps310/advanced.py]:
```suggestion
_DPS310_DEFAULT_ADDRESS = const(0x77)  # DPS310 default i2c address
#_DPS310_DEVICE_ID = const(0x10)  # DPS310 device identifier

#_DPS310_PRSB2 = const(0x00)  # Highest byte of pressure data
#_DPS310_TMPB2 = const(0x03)  # Highest byte of temperature data
_DPS310_PRSCFG = const(0x06)  # Pressure configuration
_DPS310_TMPCFG = const(0x07)  # Temperature configuration
#_DPS310_MEASCFG = const(0x08)  # Sensor configuration
#_DPS310_CFGREG = const(0x09)  # Interrupt/FIFO configuration
#_DPS310_RESET = const(0x0C)  # Soft reset
#_DPS310_PRODREVID = const(0x0D)  # Register that contains the part ID
#_DPS310_TMPCOEFSRCE = const(0x28)  # Temperature calibration src
```
Most of these constants aren't used in `advanced.py`. You could either leave them as comments in case they might be needed later or remove them as they are in `basic.py`

File: adafruit_dps310/advanced.py
Comment (lesamouraipourpre) [inline adafruit_dps310/advanced.py]:
```suggestion
```
This is called in the `basic.py` `__init__`.

File: adafruit_dps310/basic.py
Comment (lesamouraipourpre) [inline adafruit_dps310/basic.py]:
```suggestion
            from adafruit_dps310.basic import DPS310
```

File: adafruit_dps310/basic.py
Comment (lesamouraipourpre) [inline adafruit_dps310/basic.py]:
```suggestion
            dps310 = DPS310(i2c)
```

**Review by lesamouraipourpre — CHANGES_REQUESTED** [review]
Also, please consider adding `sea_level_pressure` as a property to `basic.py`. It is referenced in multiple places but there is no getter/setter for it.

For the elevation property, please add some documentation stating that the sea_level_pressure needs to be set accurately and daily (or more frequently if the pressure is changing quickly) to give accurate results. If they are only using the pressure and not the elevation then this does not need set. Unfortunately, I do not know of a good source of sea-level pressure values to recommend.
_*Note:*_ My Garmin feinx 6 can drift by almost 100m of elevation (usually less than 50m) if I have not used GPS to record an activity or manually calibrate it in a while.

Circup
--------
There should also be instructions explaining that `circup` may fail on a small board (as it will try to install everything) and explain how to download the library from github/PyPI and only install the `basic.py`.

File: README.rst
Comment (lesamouraipourpre) [inline README.rst]: The code that was removed here is tied into the explanation before it. It either needs restored or the Caveat needs removed as well.

**Comment by jposada202020** [comment]
@lesamouraipourpre. Thanks for all the recommendation. Just to clarify two points

> For the elevation property, please add some documentation stating that the sea_level_pressure needs to be set accurately and daily (or more frequently if the pressure is changing quickly) to give accurate results. If they are only using the pressure and not the elevation then this does not need set. Unfortunately, I do not know of a good source of sea-level pressure values to recommend.

This is a good point and it valid to all the `altitude` sensors, however, the scope of this PR was not intended for this. My preference is to keep PR as granular as possible to track changes and revert them if they are not working.

> Also, please consider adding `sea_level_pressure` as a property to `basic.py`. It is referenced in multiple places but there is no getter/setter for it.

If I am not mistaken the altitude property is in basic around line 222.

**Comment by jposada202020** [comment]
I will put this as Draft, as I will retest after all the suggestions included

File: adafruit_dps310/basic.py
Comment (lesamouraipourpre) [inline adafruit_dps310/basic.py]:
```suggestion
    Depending on your board memory availability you could use DPS310_Advanced.
```

**Review by lesamouraipourpre — CHANGES_REQUESTED** [review]
Sorry, I missed a couple on the last pass.

Also, the docs aren't building because of my suggested changes, `conf.py` needs updated.
```diff
diff --git a/docs/conf.py b/docs/conf.py
index feb5500..40e45fc 100644
--- a/docs/conf.py
+++ b/docs/conf.py
@@ -25,7 +25,7 @@ extensions = [
 # Uncomment the below if you use native CircuitPython modules such as
 # digitalio, micropython and busio. List the modules you use. Without it, the
 # autodoc module docs will fail to generate with a warning.
-autodoc_mock_imports = ["adafruit_register", "adafruit_bus_device"]
+autodoc_mock_imports = ["adafruit_register", "adafruit_bus_device", "micropython"]
 
 
 intersphinx_mapping = {
```

File: README.rst
Comment (jposada202020) [reply README.rst]: Oh I see Gotcha, will do

**Comment by lesamouraipourpre** [comment]
> @lesamouraipourpre. Thanks for all the recommendation. Just to clarify two points
> ...
> If I am not mistaken the altitude property is in basic around line 222.

It was the sea_level_pressure property I was looking to have added.

**Comment by jposada202020** [comment]
@lesamouraipourpre I will work on this after finish the PR on the BME280, working in two projects at the same time,  but thank you for all these suggestions I think I have something for this weekend ready.

**Comment by jposada202020** [comment]
> > @lesamouraipourpre. Thanks for all the recommendation. Just to clarify two points
> > ...
> > If I am not mistaken the altitude property is in basic around line 222.
> 
> It was the sea_level_pressure property I was looking to have added.
I see, I commit your suggestion, let me know if this is enough or maybe we could include it in the class _init__

**Comment by lesamouraipourpre** [comment]
> > > @lesamouraipourpre. Thanks for all the recommendation. Just to clarify two points
> > > ...
> > > If I am not mistaken the altitude property is in basic around line 222.
> > 
> > 
> > It was the sea_level_pressure property I was looking to have added.
>  I see, I commit your suggestion, let me know if this is enough or maybe we could include it in the class _init__

I was thinking of something like:
```diff
diff --git a/adafruit_dps310/basic.py b/adafruit_dps310/basic.py
index 375da32..a8e7811 100644
--- a/adafruit_dps310/basic.py
+++ b/adafruit_dps310/basic.py
@@ -146,8 +146,8 @@ class DPS310:
             1040384,
             2088960,
         )
-        self.sea_level_pressure = 1013.25
-        """Pressure in hectoPascals at sea level. Used to calibrate :attr:`altitude`."""
+        self._sea_level_pressure = 1013.25
+
         self.initialize()
 
     def initialize(self):
@@ -198,7 +198,7 @@ class DPS310:
 
     @property
     def pressure(self):
-        """Returns the current pressure reading in kPA"""
+        """Returns the current pressure reading in hPA"""
 
         temp_reading = self._raw_temperature
         raw_temperature = self._twos_complement(temp_reading, 24)
@@ -223,7 +223,20 @@ class DPS310:
     def altitude(self):
         """The altitude based on the sea level pressure (:attr:`sea_level_pressure`) -
         which you must enter ahead of time)"""
-        return 44330 * (1.0 - math.pow(self.pressure / self.sea_level_pressure, 0.1903))
+        return 44330 * (
+            1.0 - math.pow(self.pressure / self._sea_level_pressure, 0.1903)
+        )
+
+    @property
+    def sea_level_pressure(self):
+        """The local sea level pressure in hectoPascals (aka millibars). This is used
+        for calculation of :py:attr:`altitude`. Values are typically in the range
+        980 - 1030."""
+        return self._sea_level_pressure
+
+    @sea_level_pressure.setter
+    def sea_level_pressure(self, value):
+        self._sea_level_pressure = value
 
     @property
     def temperature(self):
```

**Comment by jposada202020** [comment]
I see, let see if it fits in the basic running in a QtPy, but I like it very much, hopefully it will fit :)

**Comment by jposada202020** [comment]
@lesamouraipourpre the sea_level_pressure property fits, so I added. Could you help me with a fgood explanation on the Readme regarding how to use the library.
We can install the library, and use it with Circup or Pypi, however for small boards we must use the `mpy` version of the `basic.py` file, the other will raise a memory error.
But you could have them install it in the board no problem.
```python
Adafruit CircuitPython 6.2.0 on 2021-04-05; Adafruit QT Py M0 with samd21e18
>>> import dps310_simpletest
Temperature = 27.15 *C
Pressure = 1015.92 hPa
Sea Level Pressure = 1013.25 
```

**Review by lesamouraipourpre — APPROVED** [review]
I've tested the changes and they work fine on a QT-PY (with only the basic installed).

Attached is diff against README.rst with a possible explanation of how to install on memory-limited boards.
[dps310.diff.txt](https://github.com/adafruit/Adafruit_CircuitPython_DPS310/files/6556530/dps310.diff.txt)

**Comment by jposada202020** [comment]
Oh..... I love the comments and the explanation, make me laugh, it was very good Thanks!!
I think we will need to get @kattni approval.
For me is just perfect :)

**Review by kattni — CHANGES_REQUESTED** [review]
Thanks everyone for contributing to this! For the README update, I viewed the diff, and there are some minor issues with it that can be addressed once it's included in this or a future PR. Please decide how you want to handle that, and get it added here if that's the plan so I can request changes.

I also want to make sure folks are aware that these changes affect at least one guide. @jposada202020 are you prepared to update the guide, or did you have a plan for that? I can't remember if we got you added to the Learn system or not.

**Comment by jposada202020** [comment]
@kattni Yes I have been added. more details on that..
I propose that we made the changes on the readme after this PR is merged.
I also propose to merge this PR when I am being able 100% to do the learning guide.
There is no rush and I prefer that both things are in sync, hope you do not mind @lesamouraipourpre . This is an old issue, and I consider this more as an improvement, so no need to rush on my part.

**Comment by jposada202020** [comment]
@lesamouraipourpre While looking at the learning guide to make the changes, something came up I would like to test this with Blinka, (both in the RP and the Pico with Micropython). what do you think?
just to check those boxes too :)

**Review by kattni — APPROVED** [review]
As long as the guide is updated in tandem, I'm prepared to approve this. Please try to remember to tag me on the PR for the README update.

Thanks!

**Comment by jposada202020** [comment]
Will  do thanks

**Comment by lesamouraipourpre** [comment]
> 
> 
> @lesamouraipourpre While looking at the learning guide to make the changes, something came up I would like to test this with Blinka, (both in the RP and the Pico with Micropython). what do you think?
> just to check those boxes too :)

If I remember tomorrow I can try it on my RPi3 with the new library, seeing as that's its normal home for logging to adafruit.io.
I also have some Picos about but haven't tried them with Blinka/Micropython yet.

**Comment by lesamouraipourpre** [comment]
RaspberryPi 3

All examples run correctly.
I've modified and simplified my monitoring code to the following which transmits to Adafruit.IO correctly as well:
```python
import board
import busio
import time

# pip install adafruit-io
from Adafruit_IO import *

from adafruit_dps310.advanced import DPS310_Advanced as DPS310

def configure_dps310_sensor(i2c):
    dps310 = DPS310(i2c)

    dps310.wait_temperature_ready()
    dps310.wait_pressure_ready()

    # Use the figure from your local forecast for today.
    # hPa (hectoPascals) == millibars
    dps310.sea_level_pressure = 1015.0
    print("Altitude: {:.0f}m".format(dps310.altitude))

    return dps310

def configure_adafruit_io():
    try:
        from secrets import secrets
    except ImportError:
        print ("secrets are kept in secrets.py, please add them there!")
        raise

    return Client(secrets["aio_username"], secrets["aio_key"])

def main():
    i2c = board.I2C()
    # i2c = busio.I2C(board.SCL, board.SDA)

    room_sensor = configure_dps310_sensor(i2c)
    io = configure_adafruit_io()

    while True:
        temperature = room_sensor.temperature
        pressure = room_sensor.pressure
        print ("{:.1f}° {:.1f}".format(temperature, pressure))

        try:
            io.send_data("room-temperature", temperature)
            io.send_data("room-pressure", pressure)
        except Exception:
            # IOError, so wait longer
            time.sleep(300)

        time.sleep(30)

if __name__ == "__main__":
    main()
```

**Comment by jposada202020** [comment]
That is a nice example, would you mind if we use it? for the library?

**Comment by lesamouraipourpre** [comment]
> That is a nice example, would you mind if we use it? for the library?

Go for it. License is Unlicense to keep `reuse` happy. And maybe add a link to an Adafruit.IO learn guide.
I've not tried this on a microcontroller, I've got a nagging memory that it might be Blinka only but I can't remember why.

example secrets.py:
```python
secrets = {
    "aio_username" : "YOUR_USERNAME",
    "aio_key" : "YOUR_AIO_TOKEN",
}
```

**Comment by jposada202020** [comment]
Will do in a different PR, thanks

**Comment by lesamouraipourpre** [comment]
Micropython on RP Pico is a no-go. There seem's to be an issue with `Adafruit_CircuitPython_Register` getting the wrong value; if I bypass Register and manually read the I2C bus I get the correct value.

@tannewt Is there any reason that you can think of that Register wouldn't work on MicroPython. If you think it should work, I'll work up minimal example code to file

**Comment by jposada202020** [comment]
I found the same problem,  Did not go as deep as yourself in the troubleshooting. See issue here https://github.com/adafruit/Adafruit_Blinka/issues/482.
But I think this new info is important for @makermelissa

**Comment by makermelissa** [comment]
Thanks for finding that @lesamouraipourpre and @jposada202020. I'm working on finishing up a guide, but am curious to look more into that right after. I suspect register is using some feature of Blinka that's perhaps implemented differently. If either of you want to take a deeper look, that would be awesome.

**Comment by lesamouraipourpre** [comment]
Using the pending PR on Blinka (https://github.com/adafruit/Adafruit_Blinka/pull/483) this library works on Blinka/MicroPython on RP2040.

**Comment by jposada202020** [comment]
That is good news, I think I will merge over the weekend, that way when releasing next week I could correct the learning guide.
Thanks for checking that out

**Comment by FoamyGuy** [comment]
I merged main and resolved conflicts to get this passing CI. Successfully tested the basic and advanced simpletest scripts with a Feather RP2040 and DPS310.

I'm going to merge this and update the relevant parts of the guide page now

**Comment by ladyada** [comment]
thanks!

**Pull request merged**
